benchmarks/scripts/expval_benchmark.py

```diff
@@ -1,99 +1,148 @@
+            qiskit_compiled.save_density_matrix()
+            return qiskit_compiled
+
+        case "cirq":
+            cirq_compiled_qasm = cirq.qasm(cirq_compile(circuit))
+            cirq_compiled_qiskit = qasm2.loads(cirq_compiled_qasm)
+            cirq_compiled_qiskit.save_density_matrix()
+            return cirq_compiled_qiskit
+
+        case _:
+            raise ValueError(f"Unknown compiler alias: {compiler_alias}")
 
 
 def simulate_density_matrix(circuit: qiskit.QuantumCircuit) -> np.ndarray:
     depolarizing_noise = NoiseModel()
-    error = depolarizing_error(0.03, 1)
-    two_qubit_error = depolarizing_error(0.05, 2)
-    depolarizing_noise.add_all_qubit_quantum_error(error, ["u1", "u2", "u3"])
-    depolarizing_noise.add_all_qubit_quantum_error(two_qubit_error, ["cx"])
+    error = depolarizing_error(0.01, 1)
+    two_qubit_error = depolarizing_error(0.02, 2)
+    # TODO: errors should only be added to the gateset that we are compiling to
+    # but there is a bug with cirq currently compiling to U3 and CZ
+    depolarizing_noise.add_all_qubit_quantum_error(
+        error, ["u1", "u2", "u3", "rx", "ry", "rz", "h"]
+    )
+    depolarizing_noise.add_all_qubit_quantum_error(
+        two_qubit_error, ["cx", "cz"]
+    )
 
     simulator = AerSimulator(
         method="density_matrix", noise_model=depolarizing_noise
     )
     return simulator.run(circuit).result().data()["density_matrix"]
 
 
-density_matrices = {
-    compiler: simulate_density_matrix(circuit)
-    for compiler, circuit in generate_compiled_circuits(qasm_string).items()
-}
+def qiskit_gateset(circuit: qiskit.QuantumCircuit) -> set[str]:
+    everything = set(circuit.count_ops().keys())
+    return everything - {"save_density_matrix"}
+
+
+uncompiled_qiskit_circuit = get_native_rep(qasm_string, "qiskit")
+native_circuit = get_native_rep(qasm_string, compiler_alias)
+compiled_circuit = compile_for_simulation(native_circuit, compiler_alias)
+
+if log:
+    circuit_name = os.path.split(qasm_path)[-1]
+    print(f"Compiling {circuit_name} with {compiler_alias}")
+    print(
+        f"    Gate reduction: {len(uncompiled_qiskit_circuit)} -> {len(compiled_circuit) - 1}"
+    )  # minus 1 to account for the addition of `save_density_matrix`
+    print(
+        f"    Starting gate set: {qiskit_gateset(uncompiled_qiskit_circuit)}"
+    )
+    print(f"    Final gate set:    {qiskit_gateset(compiled_circuit)}")
+    print(f"    Starting gates: {uncompiled_qiskit_circuit.count_ops()}")
+    print(f"    Final gates:    {compiled_circuit.count_ops()}")
 
+density_matrix = simulate_density_matrix(compiled_circuit)
 
-observable = Operator.from_label("ZZZZZZZZZZ")
+obs_str = "Z" * compiled_circuit.num_qubits
+observable = Operator.from_label(obs_str)
 
-expectation_values = {
-    compiler: np.real(dm.expectation_value(observable))
-    for compiler, dm in density_matrices.items()
-}
+compiled_ev = np.real(density_matrix.expectation_value(observable))
```

Comment:
Asking to understand .. here and below, why do we need the `np.real`?

Reply:
Right so expectation values should always be real, but sometimes `density_matrix` isn't _exactly_ a density matrix, or maybe `observable` isn't _exactly_ an observable[^1], so it's possible that the method call returns a complex number. This means that the return type of `expectation_value` is `complex`, even though in all the cases I've seen, it returns a real number (but in the form `4.32+0.0j`). I'm not sure if it's the best way, but I'm using `np.real` to get rid of the complex part to ensure all of our errors are real.


[^1]: In the mathematical senses of the words.